## Teste de Lógica e Programação Physa

### Jogo de Tabuleiro
Você e seu amigo estão jogando o jogo de tabuleiro abaixo.

### Regras do jogo
Cada jogador começa na casa 1 e se reveza jogando um dado justo de 6 lados. O jogador se move
o número de casas indicado no dado. Se você cair na base de uma escada, você automaticamente
se move para a casa no topo da escada. Por outro lado, se você cair na cabeça de uma cobra,
então você desce para a casa na ponta do rabo da cobra. O vencedor é a primeira pessoa a chegar
ou passar pela última casa.

O tabuleiro tem 36 casas conforme abaixo:


![Tabuleiro do jogo](images/dicegame.png)



### Objetivo do teste
Estando interessado em análises, você decide fazer simulações de 10.000 jogos para entender
suas chances de ganhar sob diferentes cenários. Considere cada cenário independente dos
outros.

Por favor, responda às seguintes perguntas usando quaisquer meios disponíveis. Gostaríamos que
você mantivesse um registro do seu trabalho e nos guiasse através da sua solução. Estamos
principalmente interessados na sua abordagem e código, em vez das respostas finais.


# Soluções

## Implementação do jogo

In [62]:
import random

tabuleiro_especial = {
    # Escadas
    3: 16, 5: 7, 15: 25, 18: 20, 21: 32,
    # Cobras
    12: 2, 17: 4, 14: 11, 31: 19, 35: 22  
}

def jogar(posicao_atual):
    dado = random.randint(1, 6)
    nova_posicao = posicao_atual + dado
    print("Valor do dado: ",+dado)
    if nova_posicao in tabuleiro_especial:
        tipo = "Escada ⬆️" if tabuleiro_especial[nova_posicao] > nova_posicao else "Cobra ⬇️"
        print(f"Caiu na {tipo} na casa {nova_posicao}!")
        nova_posicao = tabuleiro_especial[nova_posicao]
        
    return nova_posicao

posicao_jogadores = [1, 1]
vez = 0
meta = 36

while posicao_jogadores[0] < meta and posicao_jogadores[1] < meta:
    print(f"\n--- Turno do Jogador {vez + 1} ---")
    
    posicao_jogadores[vez] = jogar(posicao_jogadores[vez])
    
    print(f"Posição atual: {posicao_jogadores[vez]}")
    
    if posicao_jogadores[vez] >= meta:
        print(f"\n🏆 Vitória para o jogador: {vez + 1}!")
        break
        
    vez = 1 - vez# Troca de turno



--- Turno do Jogador 1 ---
Valor do dado:  3
Posição atual: 4

--- Turno do Jogador 2 ---
Valor do dado:  5
Posição atual: 6

--- Turno do Jogador 1 ---
Valor do dado:  3
Posição atual: 7

--- Turno do Jogador 2 ---
Valor do dado:  3
Posição atual: 9

--- Turno do Jogador 1 ---
Valor do dado:  5
Caiu na Cobra ⬇️ na casa 12!
Posição atual: 2

--- Turno do Jogador 2 ---
Valor do dado:  1
Posição atual: 10

--- Turno do Jogador 1 ---
Valor do dado:  5
Posição atual: 7

--- Turno do Jogador 2 ---
Valor do dado:  5
Caiu na Escada ⬆️ na casa 15!
Posição atual: 25

--- Turno do Jogador 1 ---
Valor do dado:  4
Posição atual: 11

--- Turno do Jogador 2 ---
Valor do dado:  3
Posição atual: 28

--- Turno do Jogador 1 ---
Valor do dado:  3
Caiu na Cobra ⬇️ na casa 14!
Posição atual: 11

--- Turno do Jogador 2 ---
Valor do dado:  2
Posição atual: 30

--- Turno do Jogador 1 ---
Valor do dado:  3
Caiu na Cobra ⬇️ na casa 14!
Posição atual: 11

--- Turno do Jogador 2 ---
Valor do dado:  1
Caiu na Cob

## Pergunta 01:

Num jogo de duas pessoas, qual é a probabilidade do jogador que começa o jogo vencer?

### Implementação para questão 01:

In [80]:
import random

tabuleiro_especial = {
    # Escadas
    3: 16, 5: 7, 15: 25, 18: 20, 21: 32,
    # Cobras
    12: 2, 17: 4, 14: 11, 31: 19, 35: 22  
}

def jogar(posicao_atual):
    dado = random.randint(1, 6)
    nova_posicao = posicao_atual + dado
    if nova_posicao in tabuleiro_especial:
        nova_posicao = tabuleiro_especial[nova_posicao]
        
    return nova_posicao

meta = 36
count = 0
conta_vitoria = [0,0]

while count < 10000:
    posicao_jogadores = [1, 1]
    vez = 0

    while posicao_jogadores[0] < meta and posicao_jogadores[1] < meta:
        
        posicao_jogadores[vez] = jogar(posicao_jogadores[vez])
        
        if posicao_jogadores[vez] >= meta:
            conta_vitoria[vez] = conta_vitoria[vez]+1
        vez = 1 - vez

    count = count + 1
print(f"Jogador 01: {conta_vitoria[0]/100}%")
print(f"Jogador 02: {conta_vitoria[1]/100}%")

Jogador 01: 52.85%
Jogador 02: 47.15%


### Resposta:
A probabilidade média do jogador que **começa** vencer a partida é de 52.85%.

Já a probabilidade do jogador que joga **depois** vencer é de 47.15%.

## Pergunta 02:
Em média, em quantas cobras os jogadores caem a cada jogo?

### Implementação para questão 02:

In [81]:
import random

tabuleiro_especial = {
    # Escadas 
    3: 16, 5: 7, 22: 30, 15: 32, 19: 31,
    # Cobras 
    12: 2, 20: 4, 23: 11, 31: 13, 35: 16
}

INDICE_ESCADA = 0
INDICE_COBRA = 1

def jogar(posicao_atual, vez, contador_eventos):
    dado = random.randint(1, 6)
    nova_posicao = posicao_atual + dado
    
    if nova_posicao in tabuleiro_especial:
        destino = tabuleiro_especial[nova_posicao]
        tipo_evento = INDICE_ESCADA if destino > nova_posicao else INDICE_COBRA
        contador_eventos[vez][tipo_evento] += 1
        return destino
    
    return nova_posicao


total_simulacoes = 10000
meta = 36
vitorias = [0, 0]

total_eventos = [[0, 0], [0, 0]] 

for _ in range(total_simulacoes):
    posicao_jogadores = [1, 1]
    contador_partida = [[0, 0], [0, 0]]
    vez = 0
    
    while max(posicao_jogadores) < meta:
        posicao_jogadores[vez] = jogar(posicao_jogadores[vez], vez, contador_partida)
        
        if posicao_jogadores[vez] >= meta:
            vitorias[vez] += 1
            break 
        
        vez = 1 - vez 
    
    for p in range(2):
        total_eventos[p][0] += contador_partida[p][0]
        total_eventos[p][1] += contador_partida[p][1]


print(f"Resultados após {total_simulacoes} partidas:")
print(f"Vitórias Jogador 1: {vitorias[0] / total_simulacoes * 100:.1f}%")
print(f"Vitórias Jogador 2: {vitorias[1] / total_simulacoes * 100:.1f}%")
print("-" * 30)

media_cobras_geral = (total_eventos[0][1] + total_eventos[1][1]) / total_simulacoes
print(f"Média de cobras por jogo (geral): {media_cobras_geral:.2f}")
media_cobras_geral = (total_eventos[0][1]) / total_simulacoes
print(f"- Média de cobra por jogo (jogador 01): {media_cobras_geral:.2f}")
media_cobras_geral = (total_eventos[1][1]) / total_simulacoes
print(f"- Média de cobra por jogo (jogador 02): {media_cobras_geral:.2f}")
print()
media_cobras_geral = (total_eventos[0][0] + total_eventos[1][0]) / total_simulacoes
print(f"Média de escada por jogo (geral): {media_cobras_geral:.2f}")
print()


Resultados após 10000 partidas:
Vitórias Jogador 1: 52.5%
Vitórias Jogador 2: 47.5%
------------------------------
Média de cobras por jogo (geral): 1.90
- Média de cobra por jogo (jogador 01): 0.99
- Média de cobra por jogo (jogador 02): 0.91

Média de escada por jogo (geral): 2.67



### Resposta:
A média de **cobras** por jogo é: 1.90
 - 0.99 para o jogador 1
 - 0.91 para o jogador 2

## Pergunta 03:
Se cada vez que um jogador cair em uma escada houver apenas 50% de chance de ele poder subir, qual é o número médio de lançamentos de dados necessários para completar um jogo?

### Implementação para questão 03:

In [88]:
import random

tabuleiro_especial = {
    # Escadas
    3: 16, 5: 7, 15: 25, 18: 20, 21: 32,
    # Cobras
    12: 2, 17: 4, 14: 11, 31: 19, 35: 22  
}

def jogar(posicao_atual):
    dado = random.randint(1, 6)
    nova_posicao = posicao_atual + dado
    
    if nova_posicao in tabuleiro_especial:
        destino = tabuleiro_especial[nova_posicao]
        
        if destino > nova_posicao:
            
            if random.randint(1, 2) == 1:
                return destino  
            else:
                return nova_posicao 
        return destino      
    return nova_posicao

total_simulacoes = 10000
meta = 36
count = 0
conta_dado = [0, 0]

for _ in range(total_simulacoes):
    posicao_jogadores = [1, 1]
    vez = 0

    while posicao_jogadores[0] < meta and posicao_jogadores[1] < meta:
        
        posicao_jogadores[vez] = jogar(posicao_jogadores[vez])
        conta_dado[vez] = conta_dado[vez]+1
        
        if posicao_jogadores[vez] >= meta:
            conta_vitoria[vez] = conta_vitoria[vez]+1
        vez = 1 - vez

media = (conta_dado[0] + conta_dado[1]) / total_simulacoes
print(f"A média de rolagem de dados total é de {media:.2f} vezes")
media = conta_dado[0] / total_simulacoes
print(f"A média de rolagem de dados para quem começa é de {media:.2f} vezes")
media = conta_dado[1] / total_simulacoes
print(f"A média de rolagem de dados para quem joga depois é de {media:.2f} vezes")

A média de rolagem de dados total é de 22.46 vezes
A média de rolagem de dados para quem começa é de 11.49 vezes
A média de rolagem de dados para quem joga depois é de 10.97 vezes


### Resposta:
Em média, são necessários 22.46 lançamentos para totais para concluir uma partida.
Um jogador joga em média 11 vezes para concluir a partida.

## Pergunta 04:
Você decide fazer experimentos para que os jogadores tenham aproximadamente chances iguais de vencerem o jogo. Você faz isso alterando a casa em que o Jogador 2 começa. Qual casa deveria ser a nova posição inicial do Jogador 2 para que as chances de vitória dos jogadores sejam aproximadamente iguais?

### Implementação para questão 04:

In [118]:
import random

tabuleiro_especial = {
    # Escadas
    3: 16, 5: 7, 15: 25, 18: 20, 21: 32,
    # Cobras
    12: 2, 17: 4, 14: 11, 31: 19, 35: 22  
}

def jogar(posicao_atual):
    dado = random.randint(1, 6)
    nova_posicao = posicao_atual + dado
    if nova_posicao in tabuleiro_especial:
        nova_posicao = tabuleiro_especial[nova_posicao]
        
    return nova_posicao
for j in range(1, 5):
    print(f"Execução {j}")
    print("--------------------------------------------------------------------------")
    for i in range (1, 11):
        meta = 36
        count = 0
        conta_vitoria = [0, 0]
        
        while count < 10000:
            posicao_jogadores = [1, i]
            vez = 0
        
            while posicao_jogadores[0] < meta and posicao_jogadores[1] < meta:
                
                posicao_jogadores[vez] = jogar(posicao_jogadores[vez])
                
                if posicao_jogadores[vez] >= meta:
                    conta_vitoria[vez] = conta_vitoria[vez]+1
                vez = 1 - vez
        
            count = count + 1
        print(f"Média para o jogador 02 começando na casa {i}")
        print(f"-Jogador 01: {conta_vitoria[0]/100}%")
        print(f"-Jogador 02: {conta_vitoria[1]/100}%")
        print(f"Diferença: {abs(conta_vitoria[0]-conta_vitoria[1])/100}")
        print()
    

Execução 1
--------------------------------------------------------------------------
Média para o jogador 02 começando na casa 1
-Jogador 01: 51.89%
-Jogador 02: 48.11%
Diferença: 3.78

Média para o jogador 02 começando na casa 2
-Jogador 01: 51.97%
-Jogador 02: 48.03%
Diferença: 3.94

Média para o jogador 02 começando na casa 3
-Jogador 01: 54.33%
-Jogador 02: 45.67%
Diferença: 8.66

Média para o jogador 02 começando na casa 4
-Jogador 01: 52.99%
-Jogador 02: 47.01%
Diferença: 5.98

Média para o jogador 02 começando na casa 5
-Jogador 01: 51.99%
-Jogador 02: 48.01%
Diferença: 3.98

Média para o jogador 02 começando na casa 6
-Jogador 01: 51.44%
-Jogador 02: 48.56%
Diferença: 2.88

Média para o jogador 02 começando na casa 7
-Jogador 01: 49.93%
-Jogador 02: 50.07%
Diferença: 0.14

Média para o jogador 02 começando na casa 8
-Jogador 01: 49.62%
-Jogador 02: 50.38%
Diferença: 0.76

Média para o jogador 02 começando na casa 9
-Jogador 01: 44.24%
-Jogador 02: 55.76%
Diferença: 11.52

Médi

### Resposta
Considerando os testes executados, a casa 7 é onde o jogo alcança maior equilíbrio estatístico.
Antes dessa casa, o primeiro jogador ainda possui uma vantagem consistente, especialmente ao começar na casa 3, onde ele perde a possibilidade de ganhar um grande avanço(considerando que na lógica implementada, o búnus só é ativado após a rolagem do dado). Após a casa 7, o jogador 2 passa a, consistentemente, adquirir vantagem sobre o primeiro jogador.

### Teste complementar após observações

In [121]:
import random

tabuleiro_especial = {
    # Escadas
    3: 16, 5: 7, 15: 25, 18: 20, 21: 32,
    # Cobras
    12: 2, 17: 4, 14: 11, 31: 19, 35: 22  
}

def jogar(posicao_atual):
    dado = random.randint(1, 6)
    nova_posicao = posicao_atual + dado
    if nova_posicao in tabuleiro_especial:
        nova_posicao = tabuleiro_especial[nova_posicao]
        
    return nova_posicao


meta = 36
for i in range(0, 11):
    
    count = 0
    conta_vitoria = [0,0]
    while count < 10000:
        posicao_jogadores = [3, 2]
        vez = 0
    
        while posicao_jogadores[0] < meta and posicao_jogadores[1] < meta:
            
            posicao_jogadores[vez] = jogar(posicao_jogadores[vez])
            
            if posicao_jogadores[vez] >= meta:
                conta_vitoria[vez] = conta_vitoria[vez]+1
            vez = 1 - vez
    
        count = count + 1
    print(f"Jogador 01: {conta_vitoria[0]/100}%")
    print(f"Jogador 02: {conta_vitoria[1]/100}%")
    print(f"- Diferença: {abs(conta_vitoria[0]-conta_vitoria[1])/100}")
    print()
    

Jogador 01: 50.13%
Jogador 02: 49.87%
- Diferença: 0.26

Jogador 01: 48.37%
Jogador 02: 51.63%
- Diferença: 3.26

Jogador 01: 50.25%
Jogador 02: 49.75%
- Diferença: 0.5

Jogador 01: 49.53%
Jogador 02: 50.47%
- Diferença: 0.94

Jogador 01: 49.28%
Jogador 02: 50.72%
- Diferença: 1.44

Jogador 01: 49.27%
Jogador 02: 50.73%
- Diferença: 1.46

Jogador 01: 50.44%
Jogador 02: 49.56%
- Diferença: 0.88

Jogador 01: 49.87%
Jogador 02: 50.13%
- Diferença: 0.26

Jogador 01: 50.17%
Jogador 02: 49.83%
- Diferença: 0.34

Jogador 01: 48.64%
Jogador 02: 51.36%
- Diferença: 2.72

Jogador 01: 49.86%
Jogador 02: 50.14%
- Diferença: 0.28



### Resposta complementar
Depois que eu observei o fenômeno da casa 3, testei alternativamente uma execução onde o jogador 1 começa na casa 3 e o jogador 2 começando na casa 2.
Essa configuração se mostrou uma alternativa considerável pois os valores também tem algum equilíbrio.

## Pergunta 05:
Em uma tentativa diferente de alterar as chances do jogo, em vez de mudar a posição inicial do Jogador 2, você decide dar-lhe imunidade para a primeira cobra em que ele cair. Neste caso, qual é a probabilidade aproximada de que o Jogador 1 vença?

### Implementação para questão 05:

In [139]:
import random

tabuleiro_especial = {
    # Escadas
    3: 16, 5: 7, 15: 25, 18: 20, 21: 32,
    # Cobras
    12: 2, 17: 4, 14: 11, 31: 19, 35: 22  
}

def jogar(posicao_atual, jogador_2, imunidade):
    dado = random.randint(1, 6)
    nova_posicao = posicao_atual + dado
    
    if nova_posicao in tabuleiro_especial:
        destino = tabuleiro_especial[nova_posicao]
        if destino < nova_posicao:
            if jogador_2 and imunidade:
                imunidade = 0
                return nova_posicao, False
            
        return destino, imunidade   
    return nova_posicao, imunidade

meta = 36
count = 0
conta_vitoria = [0,0]

while count < 10000:
    posicao_jogadores = [1, 1]
    vez = 0
    imunidade = True
    
    while posicao_jogadores[0] < meta and posicao_jogadores[1] < meta:
        
        posicao_jogadores[vez], imunidade = jogar(posicao_jogadores[vez], (vez==1), imunidade)
        
        if posicao_jogadores[vez] >= meta:
            conta_vitoria[vez] = conta_vitoria[vez]+1
            break
        vez = 1 - vez

    count = count + 1
print(f"Jogador 01: {conta_vitoria[0]/100}%")
print(f"Jogador 02: {conta_vitoria[1]/100}%")

Jogador 01: 38.44%
Jogador 02: 61.56%


### Resposta
Esta solução de balanceamento não se mostrou eficiente, pois inverteu a vantagem excessivamente para o Jogador 2 (taxa de vitória do Jogador 1: 38.44%). Isso ocorre porque, como verificado anteriormente, o número médio de vezes que um jogador cai em cobras por partida é baixo (geralmente próximo de 1). Assim, ao conceder imunidade, o Jogador 2 deixa de sofrer os efeitos de quase todas as cobras que encontraria, tornando a vantagem desproporcional em relação ao benefício de começar o jogo do Jogador 1.